src/analysis/convex/minmax.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2020 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import analysis.calculus.local_extr
+import topology.algebra.affine
+
+/-!
+# Minima and maxima of convex functions
+
+We show that if a function `f : E → ℝ` is convex, then a local minimum is also
+a global minimum.
+-/
+
+universes u
+
+variables {E : Type u} [add_comm_group E] [topological_space E]
+  [module ℝ E] [topological_add_group E] [topological_vector_space ℝ E]
+  {s : set E}
+
+open set filter
+open_locale classical
+
+/--
+Helper lemma for the more general case: `is_min_on.of_is_local_min_on_of_convex_on`.
+-/
+lemma is_min_on.of_is_local_min_on_of_convex_on_Icc {f : ℝ → ℝ} {a b : ℝ}
+  (a_lt_b : a < b) (h_local_min : is_local_min_on f (Icc a b) a) (h_conv : convex_on (Icc a b) f) :
+  ∀ x ∈ Icc a b, f a ≤ f x :=
+begin
+  by_contradiction H_cont,
+  push_neg at H_cont,
+  rcases H_cont with ⟨x, ⟨x_in_I, fx_lt_fa⟩⟩,
+
+  have exists_nhd_min : ∃ z > a, ∀ y ∈ (Icc a z), f y ≥ f a,
+  { dsimp [is_local_min_on, is_min_filter] at h_local_min,
+    rw [eventually_iff_exists_mem] at h_local_min,
+    rcases h_local_min with ⟨U, U_in_nhds_within, fy_ge_fa⟩,
+    rw [nhds_within_Icc_eq_nhds_within_Ici a_lt_b,mem_nhds_within_Ici_iff_exists_Icc_subset]
+        at U_in_nhds_within,
+    rcases U_in_nhds_within with ⟨ε, ε_in_Ioi, Ioc_in_U⟩,
+    refine ⟨ε, mem_Ioi.mp ε_in_Ioi, _⟩,
+    exact λ y y_in_Ioc, fy_ge_fa y (mem_of_mem_of_subset y_in_Ioc Ioc_in_U) },
+
+  rw [mem_Icc] at x_in_I,
+  rcases x_in_I with ⟨h_ax, h_xb⟩,
+
+  have a_ne_x : a ≠ x,
+  { by_contradiction H,
+    push_neg at H,
+    rw H at fx_lt_fa,
+    exact ne_of_lt fx_lt_fa rfl },
+
+  have a_lt_x : a < x := lt_of_le_of_ne h_ax a_ne_x,
+
+  have lt_on_nhd : ∀ y ∈ Ioc a x, f y < f a,
+  { intros y y_in_Ioc,
+    rcases (convex.mem_Ioc a_lt_x).mp y_in_Ioc with ⟨ya, yx, ya_pos, yx_pos, yax, y_combo⟩,
+    have h₁ : f (ya * a + yx * x) ≤ ya * f a + yx * f x :=
+      h_conv.2 (left_mem_Icc.mpr (le_of_lt a_lt_b)) ⟨h_ax, h_xb⟩ (ya_pos)
+        (le_of_lt yx_pos) yax,
+
+    rw [y_combo] at h₁,
```

Comment:
You can probably get rid of this line by using `rfl` in the `rcases`.

Reply:
This one doesn't work, since h₁ is being defined between the rcases and this line. Instead I combined the h₁ definition with the calc block below to get something a bit cleaner.